### DISCUSSION TITLE
Javascript Authenticator

### DISCUSSION THREAD

**Discussion by mdyunusraza** [keycloak/keycloak#13227] [Q&A]
Hi,
I have upgraded my server in my Dev environment to version 18. We have a requirement of using a JavaScript authenticator. 
As per the new documentation, I have created a JAR file and copied it in /providers directory and did a kc.sh build --features=scripts.
I get the following output (the one in bold is the JAR file for JS authenicator), However in the GUI I do not see my script under authenticator.

What else is required to use JS authenticator?

root@2f50c05d580b:/opt/keycloak/keycloak-18.0.2/providers# ../bin/kc.sh show-config
Current Mode: none
Runtime Configuration:
        kc.cache =  ispn (PersistedConfigSource)
        kc.config.args =  show-config (SysPropConfigSource)
        kc.db =  dev-file (PersistedConfigSource)
        kc.features =  preview (PersistedConfigSource)
        kc.health-enabled =  false (PersistedConfigSource)
        kc.home.dir =  /opt/keycloak/keycloak-18.0.2/bin/../ (SysPropConfigSource)
        kc.http-enabled =  false (PropertiesConfigSource[source=jar:file:///opt/keycloak/keycloak-18.0.2/lib/lib/main/org.keycloak.keycloak-quarkus-server-18.0.2.jar!/META-INF/keycloak.conf])
        kc.http-relative-path =  / (PersistedConfigSource)
        kc.log-console-output =  default (PropertiesConfigSource[source=jar:file:///opt/keycloak/keycloak-18.0.2/lib/lib/main/org.keycloak.keycloak-quarkus-server-18.0.2.jar!/META-INF/keycloak.conf])
        kc.log-file =  /opt/keycloak/keycloak-18.0.2/bin/../data/log/keycloak.log (PropertiesConfigSource[source=jar:file:///opt/keycloak/keycloak-18.0.2/lib/lib/main/org.keycloak.keycloak-quarkus-server-18.0.2.jar!/META-INF/keycloak.conf])
        kc.metrics-enabled =  false (PersistedConfigSource)
        **kc.provider.file.ammauth3.jar.last-modified =  1658319279100 (PersistedConfigSource)**
        kc.quarkus-properties-enabled =  false (PersistedConfigSource)
        kc.show.config =  none (SysPropConfigSource)
        kc.version =  18.0.2 (SysPropConfigSource)

**Comment by bonnm** [comment]
Hello,
You have to create new authentication flow (or copy an existing one), then you can select your custom authenticator by "Add execution". Java 11 is required to get the script work.

**Reply by mdyunusraza** [reply]
Thanks for the reply. However, it's not listed when I do the same as you suggested. 
But my Java version is 18. Could that be the issue? Does it only work with Java 11?

# java --version
**openjdk 18-ea 2022-03-22**
OpenJDK Runtime Environment (build 18-ea+36-Ubuntu-1)
OpenJDK 64-Bit Server VM (build 18-ea+36-Ubuntu-1, mixed mode, sharing)

**Reply by bonnm** [reply]
The nashorn scripting engine was removed in newer Java versions. As far as I know, it is possible to add it as a module. But I don't know how to do that (maybe https://github.com/keycloak/keycloak/pull/11322 is a starting point...).
Official Java 17 support is planned for Keycloak 19 (https://github.com/keycloak/keycloak/issues/9971) so I would recommend to run Keycloak 18 (and older) on JDK 11...

**Reply by mdyunusraza** [reply]
Thanks. I will try tomorrow with JDK 11 and update.
